### PR TITLE
Update all reduction primitives to take a block_dim parameter

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stacked on https://github.com/NVIDIA/Fuser/pull/3541

This PR is mostly mechanical. It just update all the reduction primitives with a new parameter `block_dim`, which is just `blockDim` for a kernel without warp specialization. For a kernel with warp specialization, we use a value for `block_dim`, which is the block dimension of the compute warps.

For better performance, it is beneficial if we can know whether we are using the `blockDim` or a custom value at compile time. To achieve this, I added a new helper class:

```C++
struct DefaultBlockDim {
  const uint32_t x, y, z;
  __device__ DefaultBlockDim() : x(blockDim.x), y(blockDim.y), z(blockDim.z) {}
  __device__ operator dim3() const {
    return blockDim;
  }
};
```
When a function takes `block_dim`, its type will be a template type `BlockDimT`, which is either `DefaultBlockDim` or `dim3`, where `DefaultBlockDim` means we are using `blockDim`, and `dim3` means we are using the provided custom value.

With this PR, I would consider all our reductions and broadcasts compatible with warp specialization. I expect them to just work with warp specialization. In the future, we may want to use TMA + warp specialization for reduction and persistent kernels.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!test --pybench

**Comment by zasdfgbnm** [comment]
!test --pybench

File: runtime/block_sync_default.cu [hunk 1]
```diff
@@ -6,18 +6,34 @@
...
+template <bool aligned, typename BlockDimT>
+__forceinline__ __device__ void sync(BlockDimT block_dim) {
   if constexpr (aligned) {
     __syncthreads();
-  } else {
+  } else if constexpr (std::is_same_v<BlockDimT, DefaultBlockDim>) {
     __barrier_sync(0);
+  } else {
+    uint32_t num_threads = block_dim.x * block_dim.y * block_dim.z;
+    asm volatile("bar.sync 0, %0;" : : "r"(num_threads) : "memory");
```
Comment (zasdfgbnm) [inline runtime/block_sync_default.cu:36]: Thanks @rdspring1 for https://github.com/NVIDIA/Fuser/pull/3516#pullrequestreview-2476538412

**Review by rdspring1 — APPROVED** [review]
LGTM.

Can you remove the `GTEST_SKIP` from `TEST_P(TmaCircularBufferingTest, Persistent)` in `tests/cpp/test_circular_buffering.cpp`?

**Comment by zasdfgbnm** [comment]
> LGTM.
> 
> Can you remove the `GTEST_SKIP` from `TEST_P(TmaCircularBufferingTest, Persistent)` in `tests/cpp/test_circular_buffering.cpp`? Or is there something prohibiting this?

It is already removed: https://github.com/NVIDIA/Fuser/pull/3536/files#diff-2f39954d1c6a4358eba10c1e833a07963d4a07a67d9d29f74b6f5518630d2f7cL1489

**Comment by zasdfgbnm** [comment]
!build

File: runtime/block_sync_default.cu (hunk 1 above)
Comment (naoyam) [reply runtime/block_sync_default.cu:36]: Is this because the additional threads for warp specialization don't participate in calling `sync`?

Comment (zasdfgbnm) [reply runtime/block_sync_default.cu:36]: Yes

**Review by naoyam — APPROVED** [review]
LGTM too

**Pull request merged**